File: flux-basic-java/src/main/java/com/example/app/user/api/SetRole.java
package com.example.app.user.api;

import com.example.app.authentication.RequiresRole;
import com.example.app.authentication.Role;
import com.example.app.user.api.model.UserProfile;
import io.fluxzero.sdk.persisting.eventsourcing.Apply;
import jakarta.validation.constraints.NotNull;

@RequiresRole(Role.ADMIN)
public record SetRole(@NotNull UserId userId, Role role) implements UserUpdate {
    @Apply
    UserProfile apply(UserProfile profile) {
        return profile.toBuilder().role(role).build();
    }
}

File: flux-basic-kotlin/src/main/kotlin/com/example/app/user/api/SetRole.kt
package com.example.app.user.api

import com.example.app.authentication.RequiresRole
import com.example.app.authentication.Role
import com.example.app.user.api.model.UserProfile
import io.fluxzero.sdk.persisting.eventsourcing.Apply
import jakarta.validation.constraints.NotNull

@RequiresRole(Role.ADMIN)
data class SetRole(
    @field:NotNull override val userId: UserId,
    val role: Role?
) : UserCommand {
    @Apply
    fun apply(profile: UserProfile): UserProfile {
        return profile.toBuilder().role(role).build()
    }
}
